pilot/pkg/model/sidecar.go

```diff
@@ -0,0 +1,324 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"strings"
+
+	networking "istio.io/api/networking/v1alpha3"
+)
+
+const (
+	wildcardNamespace = "*"
+	wildcardService   = Hostname("*")
+)
+
+// SidecarScope is a wrapper over the Sidecar resource with some
+// preprocessed data to determine the list of services, virtualServices,
+// and destinationRules that are accessible to a given sidecar.
+type SidecarScope struct {
+	// The crd itself. Can be nil if we are constructing the default
+	// sidecar scope
+	Config *Config
+
+	// set of egress listeners, and their associated services.
+	// A sidecar scope should have either ingress/egress listeners or both.
+	EgressListeners []*IstioEgressListenerWrapper
+
+	// // set of ingress listeners
+	// // A sidecar scope should have either ingress/egress listeners or both.
+	// IngressListeners []*IstioIngressListenerWrapper
+
+	// Union of services imported across all egress listeners for use by CDS code.
+	// Right now, we include all the ports in these services.
+	// TODO: Trim the ports in the services to only those referred to by the
+	// egress listeners.
+	allImportedServices []*Service
+
+	// Destination rules imported across all egress listeners
+	allImportedDestinationRules map[Hostname]*Config
+
+	// VirtualServices imported across all egress listeners
+	allImportedVirtualServices []Config
+}
+
+// IstioEgressListenerWrapper is a wrapper for networking.IstioEgressListener object.
+// It has the parsed form of the hosts field, encompassing a list of services
+// per sidecar listener object.
+type IstioEgressListenerWrapper struct {
+	// The actual IstioEgressListener api object from the Config. It can be nil if
+	// this is for the default sidecar scope.
+	IstioListener *networking.IstioEgressListener
+
+	// TODO: Unix domain socket
+
+	// optional. The port on which this listener should be configured. If
+	// omitted, we infer from services imported
+	ListenerPort *Port
+
+	// parsed form of the hosts field
+	// map of namespace and services 
+	importMap map[string]Hostname
+
+	// List of services imported by this egress listener
+	// extracted from the importMap above.
+	importedServices []*Service
+}
+
+// DefaultSidecarScope is a sidecar scope object with a default catch all egress listener
+// that matches the default Istio behavior: a sidecar has listeners for all services in the mesh
+// We use this scope when the user has not set any sidecar Config for a given config namespace.
+func DefaultSidecarScopeForNamespace(ps *PushContext, configNamespace string) *SidecarScope {
+	dummyNode := Proxy{
+		ConfigNamespace: configNamespace,
+	}
+
+	defaultEgressListener := &IstioEgressListenerWrapper{
+		importMap: map[string]Hostname{wildcardNamespace: wildcardService},
+	}
+	defaultEgressListener.importedServices = defaultEgressListener.selectServices(ps.Services(&dummyNode))
```

Comment:
We donot need select here
defaultEgressListener.importedServices =ps.Services(&dummyNode)